embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt

```diff
@@ -0,0 +1,264 @@
+package io.embrace.android.embracesdk.internal.capture.activity
+
+import android.app.Application.ActivityLifecycleCallbacks
+import android.os.Build
+import io.embrace.android.embracesdk.internal.arch.schema.EmbType
+import io.embrace.android.embracesdk.internal.spans.PersistableEmbraceSpan
+import io.embrace.android.embracesdk.internal.spans.SpanService
+import io.embrace.android.embracesdk.internal.utils.VersionChecker
+import io.embrace.android.embracesdk.spans.ErrorCode
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * Observes Activity lifecycle and rendering events to create traces that model the workflow for showing an Activity on screen.
+ * Depending on the version of Android and the state of the app, the start, end, and intermediate stages of the workflow will use
+ * timestamps from different events, which affects to preciseness of the measurement.
+ *
+ * Trace Start:
+ *
+ * - On Android 10+, when [ActivityLifecycleCallbacks.onActivityPostPaused] is fired, denoting that the previous activity has completed
+ *   its [ActivityLifecycleCallbacks.onActivityPaused] callbacks and a new Activity is ready to be created.
+ * - Android 9 and lower, when [ActivityLifecycleCallbacks.onActivityPaused] is fired, denoting that the previous activity is in the
+ *   process of exiting. This will possibly result in some cleanup work of exiting the previous activity being included in the duration
+ *   of the next trace that is logged.
+ *
+ * Trace End:
+ *
+ * - Android 10+, when the Activity's first UI frame finishes rendering and is delivered to the screen
```

Comment:
The UI frame thing? That's when the activity opening ends. The start is when it comes from the background or when the previous activity is done. The former is when we can detect onCreate (so onActivityPreCreated() if possible, and just onActivityCreated() if it's not depending on API version), while the latter is onActivityPostPaused() if possible, and just onActivityPaused().

Reply:
I'm struggling to visualise what the trace will look like in a dashboard but think that's probably just because this PR is a small part of a bigger picture.

Reply:
Eliab put together a really comprehensive outline here: https://www.notion.so/embraceio/Tracing-Auto-Instrumentation-1603dd140a8641ceb9fc8f0a1c63f3ef?pvs=4

There are probably differences between this impl and what is spec'd out, but largely this is what we want to achieve